tests/test_extensions.py

```diff
@@ -0,0 +1,91 @@
+import os
+
+import pytest
+
+try:
+    from unittest.mock import patch
+except ImportError:
+    from mock import patch
+
+# TODO: use Pytest's directory used for tmpdir/basedir, not '/tmp'
+os.environ['ALPENHORN_LOG_FILE'] = '/tmp' + '/alpenhornd.log'
+
+
+from alpenhorn import extensions, acquisition, generic
+
+
+@pytest.fixture
+def fixtures():
+
+    from alpenhorn import db
+
+    db._connect()
+
+    db.database_proxy.create_tables([
+        acquisition.AcqType, acquisition.FileType, acquisition.ArchiveAcq,
+        acquisition.ArchiveFile, generic.GenericAcqInfo,
+        generic.GenericFileInfo
+    ])
+
+    yield
+
+    # cleanup
+    db.database_proxy.close()
+
+
+def test_invalid_extension():
+    # Test that invalid extension paths, or modules that are not extensions
+    # throw the approproate exceptions
+
+    with patch('alpenhorn.config.configdict', {'extensions': ['unknown_module']}):
+        with pytest.raises(ImportError):
+            extensions.load_extensions()
+
+    with patch('alpenhorn.config.configdict', {'extensions': ['alpenhorn.acquisition']}):
+        with pytest.raises(RuntimeError):
+            extensions.load_extensions()
+
+
+def test_generic_extension(fixtures):
+    # Test that extension registration works correctly for the generic extension
+
+    conf = {
+        'extensions': ['alpenhorn.generic'],
+        'acq_types': {
+            'generic': {
+                'patterns': '*.zxc',
+                'file_types': ['generic']
+            }
+        },
+        'file_types': {
+            'generic': {
+                'patterns': '*.zxc'
+            }
+        }
+    }
+
+    # Load the extensions. This should cause the acq/file info types to be registered
+    with patch('alpenhorn.config.configdict', conf):
+        extensions.load_extensions()
+
+        extensions.register_type_extensions()
+
+    # Check that we have registered every known type
+    acquisition.AcqType.check_registration()
+    acquisition.FileType.check_registration()
+
+    # Check that the correct entries have appeared
+    assert 'generic' in acquisition.AcqType._registered_acq_types
+    assert 'generic' in acquisition.FileType._registered_file_types
+    assert generic.GenericAcqInfo is acquisition.AcqType._registered_acq_types['generic']
```

Comment:
Is the question how come `_registered_acq_types` is not a list? Or how come `_registered_acq_types['generic']` is not a list? If the former, I think it probably could be. Each of the keys should be equal to the `_acq_type` of the the AcqInfo class. I mostly stored it as a dict to ease look up of by name, figuring that as there is meant to be no direct access to the dict I didn't need to worry about the key and the `_acq_type` somehow getting out of sync.

It is a bit inelegant though, so I'd be fine if you wanted to change it.